### PR TITLE
[DebugInfo][NaryReassociate] Fix missing debug location updates

### LINKED ISSUE #92537: [DebugInfo][NaryReassociate] Missing debug location updates
NaryReassociate[-L514](https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/lib/Transforms/Scalar/NaryReassociate.cpp#L514), [-L517](https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/lib/Transforms/Scalar/NaryReassociate.cpp#L517C5-L517C70): the new Add and Mul finally replace `I` without a proper debug location assigned in [`NaryReassociatePass::doOneIteration(Function &F)`](https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/lib/Transforms/Scalar/NaryReassociate.cpp#L231).

```cpp
Instruction *NaryReassociatePass::tryReassociatedBinaryOp(const SCEV *LHSExpr,
                                                          Value *RHS,
                                                          BinaryOperator *I) {
  // Look for the closest dominator LHS of I that computes LHSExpr, and replace
  // I with LHS op RHS.
  auto *LHS = findClosestMatchingDominator(LHSExpr, I);
  if (LHS == nullptr)
    return nullptr;

  Instruction *NewI = nullptr;
  switch (I->getOpcode()) {
  case Instruction::Add:
->  NewI = BinaryOperator::CreateAdd(LHS, RHS, "", I->getIterator());
    break;
  case Instruction::Mul:
->  NewI = BinaryOperator::CreateMul(LHS, RHS, "", I->getIterator());
    break;
  default:
    llvm_unreachable("Unexpected instruction.");
  }
  NewI->takeName(I);
  return NewI;
}
```

### PR DESCRIPTION
**Pull request by Apochens**
Fix #92537 .

`preserving-debugloc-add-mul.ll` tests both NaryReassociate[-L514](https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/lib/Transforms/Scalar/NaryReassociate.cpp#L514) and [-L517](https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/lib/Transforms/Scalar/NaryReassociate.cpp#L517C5-L517C70) fixes.

### CODE REVIEW

**Comment by Apochens** [comment]
@OCHyams Could you please help in reviewing this PR?

File: llvm/lib/Transforms/Scalar/NaryReassociate.cpp [hunk 1]
```diff
@@ -512,9 +512,11 @@ Instruction *NaryReassociatePass::tryReassociatedBinaryOp(const SCEV *LHSExpr,
   switch (I->getOpcode()) {
   case Instruction::Add:
     NewI = BinaryOperator::CreateAdd(LHS, RHS, "", I->getIterator());
+    NewI->setDebugLoc(I->getDebugLoc());
     break;
   case Instruction::Mul:
     NewI = BinaryOperator::CreateMul(LHS, RHS, "", I->getIterator());
+    NewI->setDebugLoc(I->getDebugLoc());
```
Comment (OCHyams) [inline llvm/lib/Transforms/Scalar/NaryReassociate.cpp:519]: style nit: we could sink this out the switch. I'll leave it up to you.

File: llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll [hunk 2]
```diff
@@ -0,0 +1,88 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt < %s -passes=nary-reassociate -S | FileCheck %s
+
+; Test that NaryReassociate's tryReassociatedBinaryOp() propagates the
+; debug location to new `add` and `mul` from the original binary operator
+; they replaced (`%3` in both `@add_reassociate` and `@mul_reassociate`).
+; Because there are more than one identical binary operators in the optimized
+; program, autogenerated checks are used to check the whole program.
```
Comment (OCHyams) [inline llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll:8]: Can we remove the !dbg attachments (and metadata definitions which will become unused) from all the instructions that do not contribute to the new debug location? That helps to keep the test targeted and easy to read, and reduces the chance it'll get accidentally incorrectly regenerated in the future.

**Review by OCHyams — APPROVED** [review]
LGTM with two tiny nits, thanks

File: llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll (hunk 2 above)
Comment (Apochens) [reply llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll]: Would the revision be ok?

File: llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll [hunk 3]
```diff
@@ -0,0 +1,65 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt < %s -passes=nary-reassociate -S | FileCheck %s
+
```
Comment (OCHyams) [inline llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll:3]:
```suggestion
; Test that NaryReassociate's tryReassociatedBinaryOp() propagates the
; debug location to new `add` and `mul` from the original binary operator
; they replaced (`%3` in both `@add_reassociate` and `@mul_reassociate`).
```
Can you please re-add your comment?

Comment (Apochens) [reply llvm/test/Transforms/NaryReassociate/preserving-debugloc-add-mul.ll:3]: Oh, I just forget that the `update_test_checks.py` would delete the comments.🥲

**Pull request merged**